Add libhipcxx to TheRock

obersteiner opened the pull request:
## Motivation

Integration of libhipcxx into theRock.

## Technical Details

Integrate libhipcxx as a submodule and configure the pipelines to execute the lit tests.

## Test Plan

As discussed in the office hours, lit is run within the testers.

## Test Result

https://github.com/ROCm/TheRock/actions/runs/20461810715?pr=2018

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

ScottTodd commented on .github/workflows/test_component.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
     timeout-minutes: 210
     container:
-      image: ${{ inputs.platform == 'linux' && 'ghcr.io/rocm/no_rocm_image_ubuntu24_04@sha256:4150afe4759d14822f0e3f8930e1124f26e11f68b5c7b91ec9a02b20b1ebbb98' || null }}
+      image: ${{ inputs.platform == 'linux' && 'ghcr.io/rocm/no_rocm_image_ubuntu24_04:users-moberste-add_libhipcxx' || null }}
```
Tagging so we don't forget - the dockerfile updates will need to be reviewed/merged as one PR/commit first, then this can pick up a newly built image. We can't merge this PR while the image points to a user image.

geomin12 replied:
let's revert back if done!

ScottTodd commented on docs/development/windows_support.md:
```diff
@@ -68,6 +68,7 @@ mainline, in open source, using MSVC, etc.).
 | math-libs (BLAS)    | [hipSOLVER](https://github.com/ROCm/hipSOLVER)                                                                           | ✅        |                                               |
 | math-libs (BLAS)    | [hipBLAS](https://github.com/ROCm/hipBLAS)                                                                               | ✅        |                                               |
 | math-libs           | [rocWMMA](https://github.com/ROCm/rocm-libraries/tree/develop/projects/rocwmma)                                          | ✅        |                                               |
+| math-libs           | [libhipcxx](https://github.com/ROCm/libhipcxx)                                  | ❌        | Unsupported                                   |
```
Why unsupported on Windows? Is support possible/planned?

ScottTodd commented on cmake/therock_amdgpu_targets.cmake:
```diff
@@ -126,37 +135,35 @@ therock_add_amdgpu_target(gfx1102 "AMD RX 7700S/Framework Laptop 16" FAMILY dgpu
     hipSPARSELt # https://github.com/ROCm/TheRock/issues/2042
     hipBLASLt # https://github.com/ROCm/TheRock/issues/1062
+    libhipcxx
 )
```
The default is a hopeful "all architectures and platforms". We evaluate deviations from that case-by-case.

See also https://github.com/ROCm/TheRock/blob/main/ROADMAP.md and https://github.com/ROCm/ROCm/discussions/4276

geomin12 commented on .github/workflows/therock_test_harness.yml:
```diff
@@ -57,7 +57,7 @@ jobs:
     runs-on: ${{ matrix.target_bundle.test_machine }}
     container:
-      image: 'ghcr.io/rocm/no_rocm_image_ubuntu24_04@sha256:4150afe4759d14822f0e3f8930e1124f26e11f68b5c7b91ec9a02b20b1ebbb98'
+      image: 'ghcr.io/rocm/no_rocm_image_ubuntu24_04:users-moberste-add_libhipcxx'
```
same here for docker file changes, don't want to forget :)

geomin12 commented on requirements-test.txt:
```diff
@@ -4,3 +4,7 @@ pytest-check==2.5.3
+
+# libhipcxx test requirement
+lit==18.1.8
+psutil
```
can we specify a version here :) to avoid version mismatches in the future

geomin12 commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx.py:
```diff
@@ -0,0 +1,124 @@
+    "cmake",
+    "..",
+    f"-DCMAKE_PREFIX_PATH={OUTPUT_ARTIFACTS_PATH}",
+    #f"-DCMAKE_HIP_COMPILER={THEROCK_BIN_PATH}/hipcc",
```
remove comments if not used??? same goes for the 3 others below

geomin12 commented on cmake/therock_amdgpu_targets.cmake:
```diff
@@ -48,10 +48,12 @@ therock_add_amdgpu_target(gfx906 "Radeon VII / MI50 CDNA" FAMILY dgpu-all gfx90X
     hipSPARSELt # https://github.com/ROCm/TheRock/issues/2042
     composable_kernel # https://github.com/ROCm/TheRock/issues/1245
     rocWMMA # https://github.com/ROCm/TheRock/issues/1944
+    libhipcxx
```
just curious, do we ever plan to enable this for these archs? if so, can we open an issue and link the progress?

or if it is never planned to get enabled, can we add a note here as to why?

same goes for below archs

geomin12 commented on math-libs/CMakeLists.txt:
```diff
@@ -329,3 +329,50 @@ if(THEROCK_ENABLE_ROCWMMA)
+  ##############################################################################
+
+  therock_cmake_subproject_declare(libhipcxx
+    # TODO: use this in case libhipcx is moved to e.g. rocm-libraries
```
also curious, will libhipcxx be moved to rocm-libs ?

marbre replied:
If there is no plan to we should eventually also update [build_tools/bump_submodules.py](https://github.com/ROCm/TheRock/blob/main/build_tools/bump_submodules.py). Could be addressed in a follow up as I assume we try to bump via Dependabot in the first place. However, this will be problematic if the API isn't stable as this will result in us not be able to bump submodules. As @ScottTodd said, DevOps has little bandwidth to support in such cases.

obersteiner commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx.py:
```diff
@@ -0,0 +1,124 @@
+if test_type == "smoke":
+    # For smoke tests, we could filter specific tests if needed
+    # For now, run all tests but could be customized later
+    pass
```
Thanks! I removed this.

obersteiner commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx.py:
```diff
@@ -0,0 +1,124 @@
+    f"-DHIP_HIPCC_EXECUTABLE={THEROCK_BIN_PATH}/hipcc",
+    "-DLIBHIPCXX_TEST_WITH_HIPRTC=ON",
+    "-GNinja",
+  ]
```
Thanks! I split it into two separate test `test_libhipcxx_hipp.py` and `test_libhipcxx_hiprtc.py`.

geomin12 commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx_hipcc.py:
```diff
@@ -0,0 +1,111 @@
+    with open(version_file, "rt") as f:
+        loaded_file = json.load(f)
+        return loaded_file["rocm-version"]
+
```
this is a good place to add here: https://github.com/ROCm/TheRock/blob/main/build_tools/github_actions/github_actions_utils.py

ScottTodd reviewed (commented):
Marking as "reviewed" to drop this out of my review queue. When this is ready (i.e. no 'users' tag for dockerfile changes etc.) please re-request review.

This might also be easier to land as a sequence of smaller PRs.

obersteiner commented:
> > @geomin12 and @ScottTodd: What are the exact requirements so that we can merge the PR? Do we need to have 100% passing tests? We currently experience that there is a small number of failing tests which might be caused by an incompatibility or an issue within TheRock. These tests are all passing on ROCm 7.1.1.
> 
> The Docker image needs to land first, than this PR need to be updated. If this adds failing tests, we can xfail them and file an issue to further investigate. Anyway, landing this should not result in our test pipeline turning red.

I finally got rid of build-essential. So we should be fine with using Ninja and using the main-line image.

obersteiner commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx_hipcc.py:
```diff
@@ -0,0 +1,111 @@
+    "-cmake-options",
+    f"-DHIP_HIPCC_EXECUTABLE={THEROCK_BIN_PATH}/hipcc",
+]
+
```
I removed the build ci script. The test ci script can also be used without the build ci script.

obersteiner commented:
@ScottTodd @geomin12: I could get rid of the build-essential (so we do not need the docker image to be changed anymore). Also hipcc tests are already green for MI300 tests. I think I also addressed all the comments so far. What are the next steps for merging the PR? If the hiprtc tests are blocking the merge we can also add them in a second PR.

ScottTodd commented on build_tools/fetch_sources.py:
```diff
@@ -411,6 +413,12 @@ def main(argv):
+        default=True,
+        action=argparse.BooleanOptionalAction,
+        help="Include math libraries that are part of ROCM",
+    )
```
IMO this should be sorted lower, maybe in the group of `--include-rocm-media` and `--include-iree-libs` (after compilers, rocm-systems, rocm-libraries, etc.)

obersteiner commented:
> Hi @ScottTodd @marbre @geomin12: Now we managed to get hiprtc passing and also fixed the Windows tests. Now only the gfx1151 test is failing which is not supported. The issue here is that since we marked the architecture as unsupported the _toolchain.cmake does not set GPU_TARGET which then leads to a failure in `enable_language(HIP)` as we rely on GPU_TARGETS to be set correctly. Why is the cmake configure even called for an unsupported architecture? Can we remove libhipcxx in that case? Apart from that we are now passing all other pipelines. So we would be happy if we could move forward with the PR and hopefully merge it soon.

I could now get around this issue by specifying a default architecture. Now we only see a pipeline-related issue in the gfx1151 case (some driver issue). This is most-likely an infrastructure/pipeline issue. From the libhipcxx perspective we should now have tests passing.

geomin12 commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx_hipcc.py:
```diff
@@ -0,0 +1,100 @@
+# Set up ROCm/HIP environment
+environ_vars["ROCM_PATH"] = str(OUTPUT_ARTIFACTS_PATH)
+environ_vars["HIP_DEVICE_LIB_PATH"] = str(
+    OUTPUT_ARTIFACTS_PATH / "lib/llvm/amdgcn/bitcode/"
```
i assume these may get propagated to windows. can we do:

```suggestion
    OUTPUT_ARTIFACTS_PATH / "lib" / "llvm" / "amdgcn" / "bitcode"
```

obersteiner replied:
I'll leave it then unchanged ok?

obersteiner replied:
@geomin: since on the Nvidia side libcudacxx is part of CCCL (together with cub and thrust), I would keep it close to these two.

obersteiner commented on build_tools/github_actions/test_executable_scripts/test_libhipcxx_hiprtc.py:
```diff
@@ -0,0 +1,96 @@
+
+
+ROCM_VERSION = load_rocm_version()
+logging.info(f"ROCm version: {ROCM_VERSION}")
```
Some tests are (partly) disabled based on the ROCm version. The reason is that some ROCm versions suffer from bugs for which there is no WAR.

obersteiner commented on math-libs/CMakeLists.txt:
```diff
@@ -326,3 +326,51 @@ if(THEROCK_ENABLE_ROCWMMA)
+    BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/libhipcxx"
+    BACKGROUND_BUILD
+    DEFAULT_GPU_TARGETS
+      gfx942  # Fixes issue with empty target list for unsupported architectures.
```
I agree. Other projects use the same strategy to circumvent this issue that's why I added this. There is already a PR that tries to fix this issue for all projects.

ScottTodd reviewed (approved):
LGTM but I'd like another set of eyes on the artifacts. Either @marbre or @stellaraccident

obersteiner commented on math-libs/artifact-libhipcxx.toml:
```diff
@@ -0,0 +1,13 @@
+[components.test."math-libs/libhipcxx/build"]
+include = [
+  "libhipcxx/**/*",
+]
```
I had a quick look and from what I see it is mainly the source files that eat up these 20MB. I do not think that we can reduce the size in a meaningful amount by filtering out additional files. The issue with filtering only the needed files is that we do need a lot of different files for building (.cpp, .h, No extension, .hpp, .in ...). One could filter out .png and .rst files but this will not result in a meaningful reduction in size.

obersteiner replied:
@marbre: Can we remove these in a follow up PR? It should be possible to remove .upstream and docs. But I would like to do this in another PR as this one constantly creates merge conflicts -> I would like to merge this asap.

marbre replied:
If worked on in a follow up, please file an issue in TheRock so this does not get lost prior to merging.

obersteiner commented on examples/clean_configure_test_project.cmake:
```diff
@@ -14,6 +14,7 @@ set(propagate_vars
   THEROCK_ENABLE_RAND
   THEROCK_ENABLE_RCCL
   THEROCK_ENABLE_ROCWMMA
+  THEROCK_ENABLE_LIBHIPCXX
```
Done.

marbre reviewed (approved):
Works for me, but as said, please file an issue before merging.

mgehre-amd commented:
On my manual nightly run, I see build failures on gfx1153 here: 
```
[libhipcxx] /__w/TheRock/TheRock/math-libs/libhipcxx/include/cuda/std/detail/libcxx/include/__config:1411:12: error: Timing-related utility APIs (e.g., chrono) are currently not supported on the current architecture by libhipcxx. To override this error and proceed with the build, please set the compile-time flag _LIBCUDACXX_ALLOW_UNSUPPORTED_ARCHITECTURE
[libhipcxx]  1411 | #          error Timing-related utility APIs (e.g., chrono) are currently not supported on the current architecture by libhipcxx. \
[libhipcxx]       |            ^
[libhipcxx] 1 error generated when compiling for gfx1153.
```
https://github.com/ROCm/TheRock/actions/runs/20923982913/job/60116813668. This will probably show up on the next nightly run on main.
